src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala

```diff
@@ -32,39 +35,48 @@ object BlockMatrix {
     val part = GridPartitioner(blockSize, lm.rows, lm.cols)
     val lmBc = sc.broadcast(lm)
     new BlockMatrix(
-    new RDD[((Int, Int), BDM[Double])](sc, Nil) {
-      override val partitioner = Some(part)
-
-      def getPartitions: Array[Partition] = Array.tabulate(part.numPartitions)(i => IntPartition(i))
-
-      def compute(split: Partition, context: TaskContext): Iterator[((Int, Int), BDM[Double])] = {
-        val (i, j) = part.blockCoordinates(split.index)
-        val iOffset = i * blockSize
-        val jOffset = j * blockSize
-        val (blockNRows, blockNCols) = part.blockDims(split.index)
-        // FIXME return slice when write supports sliced blocks
-        val b = new BDM[Double](blockNRows, blockNCols)
-        b := lmBc.value(iOffset until iOffset + blockNRows, jOffset until jOffset + blockNCols)
-        Iterator(((i, j), b))
-      }
-    },
-    blockSize, lm.rows, lm.cols)
+      new RDD[((Int, Int), BDM[Double])](sc, Nil) {
+        override val partitioner = Some(part)
+
+        def getPartitions: Array[Partition] = Array.tabulate(part.numPartitions)(i => IntPartition(i))
+
+        def compute(split: Partition, context: TaskContext): Iterator[((Int, Int), BDM[Double])] = {
+          val (i, j) = part.blockCoordinates(split.index)
+          val iOffset = i * blockSize
+          val jOffset = j * blockSize
+          val (blockNRows, blockNCols) = part.blockDims(split.index)
+          // FIXME return slice when write supports sliced blocks
+          val b = new BDM[Double](blockNRows, blockNCols)
+          b := lmBc.value(iOffset until iOffset + blockNRows, jOffset until jOffset + blockNCols)
+          Iterator(((i, j), b))
+        }
+      },
+      blockSize, lm.rows, lm.cols)
   }
 
   def from(irm: IndexedRowMatrix): M =
     from(irm, defaultBlockSize)
 
   def from(irm: IndexedRowMatrix, blockSize: Int): M =
     irm.toHailBlockMatrix(blockSize)
-  
+
+  def fromArray(sc: SparkContext, data: java.util.ArrayList[Double], nRows: Int, nCols: Int): M = {
```

Comment:
I don't think these functions belong BlockMatrix since they really produce a BDM and we already have MatrixTable.from(BDM). They could go on RichDenseMatrixDouble (without the `hc` parameter) and go away once we have a more direct function on np.array. Could you also reorder the parameters nRows, nCols, data?

I'm uncomfortable with implicit assumes about row major vs column major data. You could make this explicit with an isTranspose parameter that's set to true in your Python call since np.reval exports row major by default and Breeze is col major.

So in the end, from Python you can call MatrixTable.from(sc, RichDenseMatrixDouble.fromArray(nRows, nCols, data, isTranspose), blockSize).  [blockSize added due to comment below]

@catoverdrive how hard is a more direct numpy_to_breeze function, inverting your breeze_to_numpy?

Reply:
FYI, not your fault that this is messy! It will get cleaner once we have local matrices exposed on the Python side of Hail.

Reply:
Actually, these functions shouldn't be necessary since BDM already has these constructors (or maybe you just need one function on RichDenseMatrixDouble to deal with ArrayList versus Array).

Reply:
Is there a reason for using RichDenseMatrixDouble instead of using DenseMatrix directly?